File: cpp_merge/classified/packages/plugin-plugin-manager/src/providers/pluginKnowledgeProvider.cpp
#include "pluginKnowledgeProvider.hpp"

std::shared_ptr<Provider> pluginKnowledgeProvider = object{
    object::pair{std:("name"), std:("pluginKnowledge")}, 
    object::pair{std:("get"), [=](auto runtime, auto message, auto state = undefined) mutable
    {
        elizaLogger->info(std:("[pluginKnowledgeProvider] Fetching plugin knowledge"));
        try
        {
            auto plugins = std::async([=]() { getAllPlugins(); });
            if (plugins->get_length() == 0) {
                return object{
                    object::pair{std:("values"), object{}}, 
                    object::pair{std:("text"), std:("No plugins available in the registry.")}
                };
            }
            shared knowledge = std:("Available ElizaOS Plugins (") + plugins->get_length() + std:(" total):\
\
");
            plugins->forEach([=](auto plugin) mutable
            {
                knowledge += std:("**") + plugin->name + std:("**\
");
                if (plugin->description) {
                    knowledge += std:("Description: ") + plugin->description + std:("\
");
                }
                if (AND((plugin->tags), (plugin->tags->get_length() > 0))) {
                    knowledge += std:("Tags: ") + plugin->tags->join(std:(", ")) + std:("\
");
                }
                knowledge += std:("\
");
            }
            );
            return object{
                object::pair{std:("values"), object{}}, 
                object::pair{std:("text"), knowledge}
            };
        }
        catch (const any& error)
        {
            elizaLogger->error(std:("[pluginKnowledgeProvider] Failed to fetch plugin knowledge:"), error);
            return object{
                object::pair{std:("values"), object{}}, 
                object::pair{std:("text"), std:("Unable to fetch plugin knowledge at this time.")}
            };
        }
    }
    }
};

void Main(void)
{
}

MAIN

File: classified/packages/plugin-plugin-manager/src/providers/pluginKnowledgeProvider.ts
import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from '@elizaos/core';
import { elizaLogger } from '@elizaos/core';
import { getAllPlugins } from '../services/pluginRegistryService';

export const pluginKnowledgeProvider: Provider = {
  name: 'pluginKnowledge',

  get: async (runtime: IAgentRuntime, message: Memory, state?: State): Promise<ProviderResult> => {
    elizaLogger.info('[pluginKnowledgeProvider] Fetching plugin knowledge');

    try {
      // Fetch all plugins from the API-based registry
      const plugins = await getAllPlugins();

      if (plugins.length === 0) {
        return {
          values: {},
          text: 'No plugins available in the registry.',
        };
      }

      // Format plugin knowledge
      let knowledge = `Available ElizaOS Plugins (${plugins.length} total):\n\n`;

      plugins.forEach((plugin) => {
        knowledge += `**${plugin.name}**\n`;
        if (plugin.description) {
          knowledge += `Description: ${plugin.description}\n`;
        }
        if (plugin.tags && plugin.tags.length > 0) {
          knowledge += `Tags: ${plugin.tags.join(', ')}\n`;
        }
        // Remove features and requiredConfig - these properties don't exist on PluginMetadata
        knowledge += '\n';
      });

      return {
        values: {},
        text: knowledge,
      };
    } catch (error) {
      elizaLogger.error('[pluginKnowledgeProvider] Failed to fetch plugin knowledge:', error);
      return {
        values: {},
        text: 'Unable to fetch plugin knowledge at this time.',
      };
    }
  },
};
